substation-diagram/substation-diagram-cgmes/src/main/java/com/powsybl/substationdiagram/cgmes/CgmesSubstationLayout.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.substationdiagram.cgmes;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.powsybl.substationdiagram.layout.LayoutParameters;
+import com.powsybl.substationdiagram.layout.SubstationLayout;
+import com.powsybl.substationdiagram.model.BusCell.Direction;
+import com.powsybl.substationdiagram.model.Edge;
+import com.powsybl.substationdiagram.model.Graph;
+import com.powsybl.substationdiagram.model.Side;
+import com.powsybl.substationdiagram.model.SubstationGraph;
+
+/**
+ *
+ * @author Massimo Ferraro <massimo.ferraro@techrain.eu>
+ */
+public class CgmesSubstationLayout extends AbstractCgmesLayout implements SubstationLayout {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CgmesSubstationLayout.class);
+
+    private final SubstationGraph graph;
+
+    public CgmesSubstationLayout(SubstationGraph graph) {
+        Objects.requireNonNull(graph);
+        for (Graph vlGraph : graph.getNodes()) {
+            removeFictitiousNodes(vlGraph);
+        }
+        fixTransformersLabel = true;
+        this.graph = graph;
+    }
+
+    @Override
+    public void run(LayoutParameters layoutParam) {
+        LOG.info("Running CGMES Substation Layout on substation {}", graph.getSubstation().getId());
+        for (Graph vlGraph : graph.getNodes()) {
+            setNodeCoordinates(vlGraph);
+        }
+        for (Graph vlGraph : graph.getNodes()) {
+            vlGraph.getNodes().forEach(node -> shiftNodeCoordinates(node, layoutParam.getScaleFactor()));
+        }
+        if (layoutParam.getScaleFactor() != 1) {
+            for (Graph vlGraph : graph.getNodes()) {
+                vlGraph.getNodes().forEach(node -> scaleNodeCoordinates(node, layoutParam.getScaleFactor()));
+            }
+        }
+    }
+
+    @Override
+    public List<Double> calculatePolylineSnakeLine(LayoutParameters layoutParam, Edge edge,
+            Map<Direction, Integer> nbSnakeLinesTopBottom, Map<Side, Integer> nbSnakeLinesLeftRight,
+            Map<String, Integer> nbSnakeLinesBetween, Map<String, Integer> nbSnakeLinesBottomVL,
+            Map<String, Integer> nbSnakeLinesTopVL) {
+        return Collections.emptyList();
```

Comment:
Question:
not sure what are the inputs to this method, but ideally shouldn't we return points for drawing transformers here ? Are these data present in the DL file ?

Reply:
The other substation layout implementations assign the coordinates to the voltage levels, place the voltage levels (horizontally or vertically), and connects them. In the substation diagrams, for each transformer, there are 2 elements in the SVG, corresponding to the 2 sides of the transformer, and the 2 elements have different coordinates. This method is used to draw the lines connecting the two sides of transformers.
The cgmes substation layout shows, in the SVG, just one element, corresponding to the transformer. So there is no need, for this substation layout, to draw any line connecting the two sides of transformer.

Reply:
Ok, thank you for the explanation.
I understand that we use the transformer position from the DL file to draw the transformer in the same place for both voltage levels (so it is "duplicated" in the SVG :) ).
Are the `terminalPoints` in `CouplingDeviceDiagramData` supposed to define the position of the wires ?
Do you think in the future this should be used for voltage level layout ? (also for switches)

Reply:
You are right, there is a transformer nodes in both (possibly 3, if it is a 3 winding transformer) voltage level graphs, but just one is shown in the svg, if you use the cgmes substation layout (the 2 or 3 nodes overlaps, showing just one ... it was the best solution I found without changing deeply the implementation, e.g. having a single graph with all the substation nodes, instead of a graph with 2 voltage level graphs inside).
The terminalPoints, in the coupling device diagram data, but also the injection and three winding transformer diagram data, could be used to draw the wires for connecting the equipments. Now these data is actually not used: the svg writer uses the same method, for all the voltage level layout, for drawing the wires. In my tests, not always the data read from cgmes, associated to terminal, is very precise. Anyway the use of this data could be a possible evolution, even if it would require refactoring of the code (as said, the wire are computed by the svg writer, non by the layout)